src/boot/fwupdate.c

```diff
@@ -0,0 +1,591 @@
+                }
+
+                _cleanup_free_ char *file_buffer = NULL;
+                size_t file_size = 0;
+                err = file_read(root, dp_str, /* offset= */ 0, /* size= */ 0, &file_buffer, &file_size);
+                if (err != EFI_SUCCESS) {
+                        log_warning_status(err, "Failed to read file for '%ls': %m, skipping", (*u)->name);
+                        continue;
+                }
+
+                if (file_size < sizeof(EFI_CAPSULE_HEADER)) {
+                        log_warning("File '%ls' is too small for EFI_CAPSULE_HEADER: %zu < %zu, skipping", (*u)->name, file_size, sizeof(EFI_CAPSULE_HEADER));
+                        continue;
+                }
+                EFI_CAPSULE_HEADER *capsule = (EFI_CAPSULE_HEADER *)file_buffer;
+
+                EFI_GUID guid = (*u)->info->guid; /* Avoid unaligned access warning, since info is packed */
+                if (efi_guid_equal(&guid, MAKE_GUID_PTR(UX_CAPSULE))) {
+                        if (UINT32_MAX - capsule->HeaderSize < (uint32_t) sizeof(UX_CAPSULE_HEADER)) {
+                                log_warning("Capsule header too large to parse UX_CAPSULE_HEADER: %" PRIu32 ", skipping",
+                                            capsule->HeaderSize);
+                                continue;
+                        }
+
+                        if (capsule->HeaderSize + sizeof(UX_CAPSULE_HEADER) > file_size) {
+                                log_warning("File '%ls' is too small for UX_CAPSULE_HEADER: %zu < %" PRIu32 ", skipping",
+                                            (*u)->name, file_size, capsule->HeaderSize);
+                                continue;
+                        }
+
+                        /* struct is packed, so no alignment check is necessary */
+                        UX_CAPSULE_HEADER *hdr = (UX_CAPSULE_HEADER *)((uint8_t *)capsule + capsule->HeaderSize);
+                        uint32_t mode = 0;
+
+                        err = get_graphics_protocol_mode(image, &mode);
+                        if (err != EFI_SUCCESS)
+                                return log_error_status(err, "Failed to get graphics protocol mode for UX capsule: %m");
+                        hdr->mode = mode; /* hdr is packed so cannot take a pointer to a field, it would be unaligned */
+
+                        /* Since the mode is updated, need to recalculate the checksum */
+                        hdr->checksum = 0;
+                        hdr->checksum = capsule_checksum((uint8_t *) hdr, sizeof(UX_CAPSULE_HEADER));
+                } else if (capsule->Flags == 0) {
+#if defined(__aarch64__) || (defined(__riscv) && __riscv_xlen == 64)
+                        capsule->Flags |= (*u)->info->capsule_flags;
+#else
+                        capsule->Flags |= (*u)->info->capsule_flags |
+                                          CAPSULE_FLAGS_PERSIST_ACROSS_RESET |
+                                          CAPSULE_FLAGS_INITIATE_RESET;
```

Comment:
i wonde rif this stuff is obsolete these days. `->capsule_flags` comes directly from whatever fwupd writes into te efi vars it generates. maybe it already does that on its own these days? seems really weird to me to patch these flags in if fwupd could just do that as well on its own, like it does on those other archs

Reply:
I suspect it's a workaround from the aarch64 bringup -- but like @poettering says -- it's the wrong place for the fixup. I think removing the workaround here is probably sane.